eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -248,9 +250,15 @@ sealed trait Data
 
 case object Nothing extends Data
 
-sealed trait HasCommitments extends Data {
-  def commitments: Commitments
+trait HasAbstractCommitments extends Data {
+  def commitments: AbstractCommitments
   def channelId: ByteVector32 = commitments.channelId
+  /** HTLCs which can't be relayed to remote peer so need to be failed in origin channel */
+  def htlcsToFailInOrigin(log: LoggingAdapter): Set[Long] = Set.empty
+}
```

Comment:
I think we can (should) avoid introducing this new trait, it's weird that it's used almost nowhere.

Why can't you change `ChannelStateChanged` to `case class ChannelStateChanged(channel: ActorRef, channelId: ByteVector32, peer: ActorRef, remoteNodeId: PublicKey, previousState: State, currentState: State, commitments: AbstractCommitments) extends ChannelEvent`? And the potential new function you need to add should be added to the `AbstractCommitments` trait, unless I'm missing something.

Reply:
Maybe I'm missing something here, I'll once again look if this trait can be avoided a provide a list of reasons which I think prevents that (if found).

Reply:
`checkBrokenHtlcs` method is the main reason `HasAbstractCommitments` exists because in order to determine which HTLCs are actually broken it needs both an instance of `AbstractCommitments` and a `Data` instance which wraps it for each channel.

Originally it takes a `Seq[HasCommitments]` and applies additional rules if wrapping `HasCommitments` instance happens to be `CLOSING_DATA`. With HTLC-enabled plugins this method needs a way to:

- access various commitment object methods (such as `htlcsRemoteCommit`, `originChannels`. Currently solved by `AbstractCommitments` abstract methods).

- know if wrapping `Data` type provides any additional rules for determining which HTLCs are broken (Currently solved by `HasAbstractCommitments` trait and its `htlcsToFailInOrigin` method which makes each data type (including plugin types) responsible for providing such rules.

Other ways of solving this that I can think of:

1. Unsealing `Data` trait (for plugin types to extend it), adding `htlcsToFailInOrigin` method to it, supplying a tuple of `(AbstractCommitments, Data)` to `checkBrokenHtlcs` method (because `Data` does not necessarily has commitments yet, so `Data`s without commitments should be filtered out). At least to me it seems less elegant than introducing `HasAbstractCommitments`.

2. Changing `HasCommitments` trait to have `AbstractCommitments` instead of `Commitments`, but https://github.com/ACINQ/eclair/pull/1580#issuecomment-715229358.